### PR TITLE
Netlify edge 2

### PR DESCRIPTION
**Pull request by foxyblocks**
## What type of PR is this? (check all applicable)

- [ ] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

### CODE REVIEW

File: netlify.toml [hunk 1]
```diff
@@ -1,11 +1,12 @@
...
 [context.alpha]
 environment = { CHANNEL = "alpha" }
 
 [context.beta]
 environment = { CHANNEL = "beta" }
 
-[dev]
-framework = "#static"
+[[plugins]]
+package = "@netlify/plugin-nextjs"
```
Comment (nickytonline) [inline netlify.toml:12]: No need to add this. It will get auto installed when you deploy.

```suggestion
```

Comment (foxyblocks) [reply netlify.toml]: So this change will cause the next.js api routes to run as edge functions?

Comment (nickytonline) [reply netlify.toml]: Sorry, should have been clearer with this change. This is something Netlify handles for you which is why you don't need it in the configuration. It's not related to anything edge.

Comment (nickytonline) [reply netlify.toml]: For the actual issue you're having is it that the og-image isn't rendering or not building during the deploy? I can take a peek at this later tonight.

**Comment by nickytonline** [comment]
After our	Discord chat, I took a peek in the codebase, but I can't find where you're using the OG image based off of ascorbic/og-image.

I made a demo site, where you can see it in action. See https://github.com/nickytonline/og-test.

I don't have enough info yet, but it sounded like you were trying enable edge functions for Next.js. You don't need to for the OG image. It's a custom edge function is all that runs outside of Next.js.

cc: @bdougie in case I'm misunderstanding the issue with OG images.

**Comment by foxyblocks** [comment]
> After our Discord chat, I took a peek in the codebase, but I can't find where you're using the OG image based off of ascorbic/og-image.
> 
> I made a demo site, where you can see it in action. See https://github.com/nickytonline/og-test.
> 
> I don't have enough info yet, but it sounded like you were trying enable edge functions for Next.js. You don't need to for the OG image. It's a custom edge function is all that runs outside of Next.js.
> 
> cc: @bdougie in case I'm misunderstanding the issue with OG images.

@nickytonline 

Thanks for putting this together. It works until it comes time to load external fonts and image assets. In our [existing serverless function](https://github.com/open-sauced/insights/blob/437ec57d8a918566127d151288e04d1cfbcacb14/pages/api/user/%5Busername%5D/card.png.tsx#L21-L27) we are loading them using a relative URL. This mirrors the examples from the vercel/og docs https://vercel.com/docs/concepts/functions/edge-functions/og-image-generation/og-image-examples#using-a-custom-font and works correctly with vercel edge functions. However, when using deno and netlify edge functions it results in a network error. 

I see in the og-edge repo they have an example of loading a custom font but they do it by hardcoding a link to the asset from the package published on deno.land https://github.com/ascorbic/og-edge/blob/d533ef878801d7f808eb004f254e782ec6ba1e3c/netlify/edge-functions/custom-font.tsx 

This probably wouldn't work for us since we are not publishing our site on deno.land and I don't like the idea of hardlinking to a specific version like this. Is there a netlify recommended way to load static assets and serve them from the latest version of the current app with edge functions? I haven't found any examples online. 

cc @bdougie

**Comment by nickytonline** [comment]
> > After our Discord chat, I took a peek in the codebase, but I can't find where you're using the OG image based off of ascorbic/og-image.
> > I made a demo site, where you can see it in action. See [nickytonline/og-test](https://github.com/nickytonline/og-test).
> > I don't have enough info yet, but it sounded like you were trying enable edge functions for Next.js. You don't need to for the OG image. It's a custom edge function is all that runs outside of Next.js.
> > cc: @bdougie in case I'm misunderstanding the issue with OG images.
> 
> @nickytonline
> 
> Thanks for putting this together. It works until it comes time to load external fonts and image assets. In our [existing serverless function](https://github.com/open-sauced/insights/blob/437ec57d8a918566127d151288e04d1cfbcacb14/pages/api/user/%5Busername%5D/card.png.tsx#L21-L27) we are loading them using a relative URL. This mirrors the examples from the vercel/og docs [vercel.com/docs/concepts/functions/edge-functions/og-image-generation/og-image-examples#using-a-custom-font](https://vercel.com/docs/concepts/functions/edge-functions/og-image-generation/og-image-examples#using-a-custom-font) and works correctly with vercel edge functions. However, when using deno and netlify edge functions it results in a network error.
> 
> I see in the og-edge repo they have an example of loading a custom font but they do it by hardcoding a link to the asset from the package published on deno.land [ascorbic/og-edge@`d533ef8`/netlify/edge-functions/custom-font.tsx](https://github.com/ascorbic/og-edge/blob/d533ef878801d7f808eb004f254e782ec6ba1e3c/netlify/edge-functions/custom-font.tsx)
> 
> This probably wouldn't work for us since we are not publishing our site on deno.land and I don't like the idea of hardlinking to a specific version like this. Is there a netlify recommended way to load static assets and serve them from the latest version of the current app with edge functions? I haven't found any examples online.
> 
> cc @bdougie

Netlify uses Deno for edge functions, but when you use edge functions, you have access to all the assets of your site as it's part of your site.

**Comment by nickytonline** [comment]
@foxyblocks, I updated the repo to demonstrate that you can have a Next.js app with the custom OG image edge function. You can see the latest source at https://github.com/nickytonline/og-test and see the update at https://og-image-test2.netlify.app/og.

And as mentioned previously, any paths to assets in the Next.js site would be available to the custom edge function.

**Comment by nickytonline** [comment]
The edge function runs in Deno, so adding some VS Code settings should sort out the dependencies not resolving. The code will still  run.

Run `netlify recipes vscode` to add a settings file.

```
➜ ntl recipes vscode
? A new VS Code settings file will be created at /Users/nicktaylor/dev/work/netli
fy-ipx/.vscode/settings.json (Y/n) 
```

For the `Netlify` global, you can do a TypeScript merge declaration, but it might be worth opening an issue for that.

```
declare global {
  // deno-lint-ignore no-var
  var Netlify: {
    env: Record<string, string>;
  };
}
```

I don't think you have access to it in Deno though. You're probably better off accessing your environment variables via [Deno's environment variables](https://examples.deno.land/environment-variables). You should have access to I believe all the Netlify.env ones.

There could definitely be some DX improvements there.

**Comment by foxyblocks** [comment]
@nickytonline Ok, i've gotten an example of our dev cards deployed to netlify. 
https://github.com/foxyblocks/netlify-edge-image-test/blob/main/netlify/edge-functions/card.tsx
It can be tested at https://64c86046f7fa6823a2181a5e--netlify-edge-image-test.netlify.app/card?username=bdougie

This is a slightly more complex example so it takes a bit longer to render. I noticed that vercel edge functions will automatically cache the result between requests (this is some sort of server side caching because it will also cache it between requests from different clients).  This caching is kind of important for the project. It doesn't look like netlify edge functions are doing anything like that. **Is there some way to get Netlify to cache the result of edge functions?** 

Also the the `netlify recipes vscode` didn't seem  to have an effect. I have the Deno extension installed, is there something else I would need?

**Comment by nickytonline** [comment]
> @nickytonline Ok, i've gotten an example of our dev cards deployed to netlify. [foxyblocks/netlify-edge-image-test@`main`/netlify/edge-functions/card.tsx](https://github.com/foxyblocks/netlify-edge-image-test/blob/main/netlify/edge-functions/card.tsx?rgh-link-date=2023-08-01T18%3A16%3A20Z) It can be tested at [64c86046f7fa6823a2181a5e--netlify-edge-image-test.netlify.app/card?username=bdougie](https://64c86046f7fa6823a2181a5e--netlify-edge-image-test.netlify.app/card?username=bdougie)
> 
> This is a slightly more complex example so it takes a bit longer to render. I noticed that vercel edge functions will automatically cache the result between requests (this is some sort of server side caching because it will also cache it between requests from different clients). This caching is kind of important for the project. It doesn't look like netlify edge functions are doing anything like that. **Is there some way to get Netlify to cache the result of edge functions?**
> 

You can cache the OG images if you add `cache: manual` to the edge function configuration. See https://docs.netlify.com/edge-functions/optional-configuration/#opt-in-to-caching

> Also the the `netlify recipes vscode` didn't seem to have an effect. I have the Deno extension installed, is there something else I would need?

Hmm, not sure what's up there. 🤔

**Comment by brandonroberts** [comment]
Closed via https://github.com/open-sauced/insights/pull/1400

**Pull request closed without merging**